pilot/pkg/proxy/envoy/config.go

```diff
@@ -134,23 +134,33 @@ func BuildConfig(config meshconfig.ProxyConfig, pilotSAN []string) *Config {
 
 // buildListeners produces a list of listeners and referenced clusters for all proxies
 func buildListeners(env model.Environment, node model.Node) (Listeners, error) {
+	var listeners Listeners
+
 	switch node.Type {
-	case model.Sidecar, model.Router:
+	case model.Sidecar:
 		instances, err := env.HostInstances(map[string]*model.Node{node.IPAddress: &node})
 		if err != nil {
 			return nil, err
 		}
+
 		services, err := env.Services()
 		if err != nil {
 			return nil, err
 		}
-		listeners, _ := buildSidecarListenersClusters(env.Mesh, instances,
-			services, env.ManagementPorts(node.IPAddress), node, env.IstioConfigStore)
-		return listeners, nil
+
+		listeners, _ = buildSidecarListenersClusters(env.Mesh, instances, services, env.ManagementPorts(node.IPAddress), node, env.IstioConfigStore)
+	case model.Router:
+		instances, err := env.HostInstances(map[string]*model.Node{node.IPAddress: &node})
+		if err != nil {
+			return nil, err
+		}
+
+		listeners = buildGatewayListeners(env.Mesh, instances, env.IstioConfigStore, node)
```

Comment:
you might want to take a look at @ijsnellf and @GregHanson 's external services PR. There is a concept of generic buildCluster function signature that can be implemented by different systems. I was telling them that we might want to follow this logic in other places as well. Thoughts? (see #2811 )